File: src/graphs1/scc.cpp
// scc.cpp — CLRS 4e §20.5：兩趟 DFS 求強連通元件（G 一趟、Gᵀ 按完成時間反序一趟）
#include <iostream>
#include <vector>

int n;
std::vector<std::vector<int>> adj, radj;
std::vector<int> visited, order, comp;

void dfs1(int u) {
    visited[u] = 1;
    for (int v : adj[u]) if (!visited[v]) dfs1(v);
    order.push_back(u);                          // 完成時間遞增
}
void dfs2(int u, int c) {
    comp[u] = c;
    for (int v : radj[u]) if (comp[v] < 0) dfs2(v, c);
}

int main() {
    // CLRS 圖 20.9：a..h = 0..7
    n = 8;
    adj.assign(n, {}); radj.assign(n, {});
    int E[][2] = {{0,1},{1,2},{1,4},{1,5},{2,3},{2,6},{3,2},{3,7},
                  {4,0},{4,5},{5,6},{6,5},{6,7},{7,7}};
    for (auto [u, v] : E) { adj[u].push_back(v); radj[v].push_back(u); }

    visited.assign(n, 0);
    for (int u = 0; u < n; ++u) if (!visited[u]) dfs1(u);

    comp.assign(n, -1);
    int c = 0;
    for (auto it = order.rbegin(); it != order.rend(); ++it)   // 完成時間遞減
        if (comp[*it] < 0) dfs2(*it, c++);

    std::cout << "共 " << c << " 個 SCC：\n";
    for (int k = 0; k < c; ++k) {
        std::cout << "  { ";
        for (int v = 0; v < n; ++v)
            if (comp[v] == k) std::cout << char('a' + v) << ' ';
        std::cout << "}\n";
    }
    return 0;
}

File: src/graphs1/scc.py
# 強連通元件（改寫自官方 Chapter 20/strongly_connected_components.py，MIT License）
# 兩趟 DFS：G 記完成順序 → G 轉置 → 按完成時間遞減收割
G = {"a": ["b"], "b": ["c", "e", "f"], "c": ["d", "g"], "d": ["c", "h"],
     "e": ["a", "f"], "f": ["g"], "g": ["f", "h"], "h": ["h"]}   # CLRS 圖 20.9

order = []                                  # 第一趟：完成順序
visited = set()


def dfs1(u):
    visited.add(u)
    for v in G[u]:
        if v not in visited:
            dfs1(v)
    order.append(u)                         # 完成時 push → order 尾端 f 最大


for u in G:
    if u not in visited:
        dfs1(u)

GT = {u: [] for u in G}                     # 轉置圖：所有邊反向
for u in G:
    for v in G[u]:
        GT[v].append(u)

comp = {}                                   # 第二趟：在 G^T 上按 f 遞減收割


def dfs2(u, c):
    comp[u] = c
    for v in GT[u]:
        if v not in comp:
            dfs2(v, c)


sccs = []
for u in reversed(order):                   # f 最大者優先當起點
    if u not in comp:
        c = len(sccs)
        dfs2(u, c)
        sccs.append([v for v in comp if comp[v] == c])
print("共 %d 個 SCC：" % len(sccs))
for s in sccs:
    print("  {", " ".join(sorted(s)), "}")
